### PR TITLE
Ignore authorization headers if not Bearer

### LINKED ISSUE #318: If `authorization` is not of type Bearer to ignore it and check the cookie
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

I think it would help for the authorization header to be skipped if it is not Bearer type. No reason to throw an error if it is not the expected format.  
Easies way to do it is to change this line 
`if ((request.headers && request.headers.authorization) && (!onlyCookie))` 
into 
`if ((request.headers && request.headers.authorization) && (!onlyCookie) && (/^Bearer/i.test(request.headers.authorization)))`

### Motivation

A good example is if you are using also the Basic auth on the same domain. 

### Example

_No response_

### PR DESCRIPTION
**Pull request by cberescu**
Fixes: https://github.com/fastify/fastify-jwt/issues/318

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Screenshot 2024-01-05 155257

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
I'm a bit conflicted if this is major or not. Anybody has an opinion?

File: test/jwt.test.js [hunk 1]
```diff
@@ -1617,20 +1617,19 @@ test('errors', function (t) {
...
-      t.test('authorization header format error', function (t) {
+      t.test('no bearer authorization header error', function (t) {
         t.plan(2)
 
         fastify.inject({
           method: 'get',
           url: '/verify',
           headers: {
-            authorization: 'Invalid Format'
+            authorization: 'Another Format'
```
Comment (climba03003) [inline test/jwt.test.js:1627]: Keep it as `Invalid` as fine.

```suggestion
            authorization: 'Invalid Format'
```

File: test/jwt.test.js [hunk 2]
```diff
@@ -2390,12 +2389,12 @@ test('custom response messages', function (t) {
...
```
Comment (climba03003) [inline test/jwt.test.js:2392]: Keep it as `Invalid` as fine.

```suggestion
            authorization: 'Invalid Format'
```

**Comment by climba03003** [comment]
> I'm a bit conflicted if this is major or not. Anybody has an opinion?

Although it stills throw Error, but changing the return from `400` to `401` status code is absolutely a major.

File: jwt.js [hunk 3]
```diff
@@ -240,15 +240,10 @@ function fastifyJwt (fastify, options, next) {
       if (!token) {
         throw new BadRequestError()
       }
-    } else if ((request.headers && request.headers.authorization) && (!onlyCookie)) {
+    } else if ((request.headers && request.headers.authorization && /^Bearer/i.test(request.headers.authorization)) && (!onlyCookie)) {
```
Comment (gurgunday) [inline jwt.js:243]: There should be space after Bearer:

```js
// Currently
/^Bearer/i.test("BearerWRONG") // true
```

```suggestion
    } else if (request.headers?.authorization && /^Bearer\s/i.test(request.headers.authorization) && !onlyCookie) {
```

Comment (gurgunday) [reply jwt.js:243]: And maybe we can extract the RegExp Object to a `const` to further increase performance

Comment (gurgunday) [reply jwt.js:243]: I see that we do that check below

Resolving

Comment (cberescu) [reply jwt.js:243]: I did not put the space as BadRequestError is more accurate than NoAuthorizationInHeaderError in a case of `BearerWRONG` . However i will add the space  if you think it is better. just let me know

Comment (gurgunday) [reply jwt.js:243]: We simply need that space, because currently, this header passes your checks:

`BearerINVALID <token>`

**Review by gurgunday — CHANGES_REQUESTED** [review]
https://github.com/fastify/fastify-jwt/pull/325#discussion_r1443742079

File: jwt.js (hunk 3 above)
Comment (cberescu) [reply jwt.js]: Good point, didn't think about it. Commited. 

Thanks!

**Review by gurgunday — APPROVED** [review]
looks good to me

File: jwt.js [hunk 4]
```diff
@@ -240,15 +240,10 @@ function fastifyJwt (fastify, options, next) {
       if (!token) {
         throw new BadRequestError()
       }
-    } else if ((request.headers && request.headers.authorization) && (!onlyCookie)) {
+    } else if (request.headers?.authorization && /^Bearer\s/i.test(request.headers.authorization) && !onlyCookie) {
```
Comment (gurgunday) [inline jwt.js:243]: Unrelated but affects performance: why don't we place `onlyCookie` in the beginning?

```suggestion
    } else if (!onlyCookie && request.headers?.authorization && /^Bearer\s/i.test(request.headers.authorization)) {
```

Comment (gurgunday) [reply jwt.js:243]: @cberescu can you do that as well?

If we set `onlyCookie = true`, we don't want to keep calling `RegExp.prototype.test` if someone for some reason repeatedly sends the auth header

File: jwt.js (hunk 3 above)
Comment (gurgunday) [reply jwt.js]: Seems like the error type changed for certain requests, can you please take a look?

File: jwt.js (hunk 4 above)
Comment (gurgunday) [reply jwt.js:243]: As long as they’re before the method call, it should be fine

> do we really need the ?.

I don’t think so

File: jwt.js (hunk 3 above)
Comment (cberescu) [reply jwt.js]: Yes, will look into it later today

**Comment by cberescu** [comment]
@gurgunday not sure why one of the tests failed... any ideas ?

**Review by gurgunday — APPROVED** [review]

**Comment by gurgunday** [comment]
> @gurgunday not sure why one of the tests failed... any ideas ?

It was a CI issue, fixed now

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by willpoorman** [comment]
Hey guys. Could you potentially update the changelog for the 8.0.0 to clarify what the exact breaking change is and what users need to adjust when upgrading? https://github.com/fastify/fastify-jwt/releases/tag/v8.0.0. I had to dig in to find this PR to read the comments to find out what was considered BREAKING about it: https://github.com/fastify/fastify-jwt/pull/325#issuecomment-1879634632

**Pull request merged**